.agents/governance/traceability-schema.md

```diff
@@ -0,0 +1,249 @@
+# Traceability Graph Schema
+
+> **Version**: 1.0.0
+> **Created**: 2025-12-31
+> **Status**: Active
+> **Related**: [enhancement-PROJECT-PLAN.md](../planning/enhancement-PROJECT-PLAN.md) Phase 2
+
+## Overview
+
+This document defines the traceability graph structure used to validate cross-references between specification artifacts (requirements, designs, and tasks).
+
+## Graph Structure
+
+### Node Types
+
+| Type | ID Pattern | Location | Description |
+|------|------------|----------|-------------|
+| REQ | `REQ-NNN` | `.agents/specs/requirements/` | EARS format requirements |
+| DESIGN | `DESIGN-NNN` | `.agents/specs/design/` | Design documents |
+| TASK | `TASK-NNN` | `.agents/specs/tasks/` | Atomic work items |
+
+### Edge Types
+
+| Edge | From | To | Meaning |
+|------|------|-----|---------|
+| `traces_to` | TASK | DESIGN | Task implements part of design |
+| `traces_to` | DESIGN | REQ | Design addresses requirement |
+| `implements` | TASK | REQ | Direct implementation reference |
+| `depends_on` | Any | Any | Prerequisite relationship |
+
+### YAML Front Matter Schema
+
+Each spec file MUST have YAML front matter with these fields:
+
+```yaml
+---
+type: requirement | design | task
+id: REQ-NNN | DESIGN-NNN | TASK-NNN
+status: draft | review | approved | implemented | complete
+related:
+  - REQ-NNN     # For designs: requirements addressed
+  - DESIGN-NNN  # For tasks: designs implemented
+---
+```
+
+## Traceability Rules
+
+### Rule 1: Forward Traceability
+
+Every REQUIREMENT must trace forward to at least one DESIGN.
+
+```text
+REQ-001 → DESIGN-001 (required)
+REQ-001 → (none)     (VIOLATION: orphaned requirement)
+```
+
+### Rule 2: Backward Traceability
+
+Every TASK must trace backward to at least one DESIGN.
+
+```text
+TASK-001 → DESIGN-001 (required)
+TASK-001 → (none)     (VIOLATION: untraced task)
+```
+
+### Rule 3: Complete Chain
+
+Every DESIGN must have both:
+- At least one REQ reference (backward)
+- At least one TASK reference (forward)
+
+```text
+REQ-001 → DESIGN-001 → TASK-001 (complete chain)
+REQ-001 → DESIGN-001 → (none)   (VIOLATION: orphaned design)
+```
+
+### Rule 4: Reference Validity
+
+All referenced IDs must exist as files.
+
+```text
+TASK-001 references DESIGN-999
+DESIGN-999 file exists?
+  YES → valid
+  NO  → VIOLATION: broken reference
+```
+
+### Rule 5: Status Consistency
+
+Completed status must propagate consistently:
+
+| If | Then |
+|-----|------|
+| TASK status = complete | DESIGN should be implemented or complete |
+| DESIGN status = implemented | REQ should be implemented |
+| All TASKs for DESIGN complete | DESIGN should be complete |
+
+## Orphan Detection Algorithm
+
+```text
+FUNCTION DetectOrphans(specs):
```

Comment:
@rjmurillo-bot This is a fun graph problem that's recomposing itself from the markdown each time. What's programming-advisor skill say about this? If we keep our own implementation of a poor man's graph database that's accessible without extra MCP (I'd like everything to be markdown and accessible without special tools, configuration, or context bloat as a project principle) then how do we make sure this is 1) super fast, 2) robust, 3) durable ?